dgl_ptm/dgl_ptm/model/initialize_model.py

```diff
@@ -407,23 +425,42 @@ def run(self):
         while self.step_count < self.step_target:
             self.step()
 
-            # save the model state every step reported by savestate
-            if self.savestate and self.step_count % self.savestate == 0:
+            # save the model state every step reported by checkpoint_period and at specific milestones.
+            # checkpoint saves overwrite the previous checkpoint; milestone get unique folders.
+            save_checkpoint = 0 < self.checkpoint_period and self.step_count % self.checkpoint_period == 0
+            save_milestone = self.milestones and self.step_count in self.milestones
+            if save_checkpoint or save_milestone:
                 self.inputs = {
                     'model_graph': copy.deepcopy(self.model_graph),
                     #'model_data': copy.deepcopy(self.model_data),
                     'generator_state': generator.get_state(),
-                    'step_count': self.step_count
+                    'step_count': self.step_count,
+                    'code_version': self.version
                 }
-                _save_model(f'./{self._model_identifier}', self.inputs)
 
+                # Note that a sinlge step could be both a checkpoint and a milestone.
+                # The checkpoint could be necessary to restore a crashed process while
+                # the milestone is required output.
+                if save_checkpoint:
+                    _save_model(f'./{self._model_identifier}', self.inputs)
+                if save_milestone:
+                    milestone_path = _make_path_unique(f'./{self._model_identifier}/milestone_{self.step_count}')
+                    _save_model(milestone_path, self.inputs)
+
+def _make_path_unique(path):
+    if Path(path).exists():
+        incr = 1
+        def add_incr(path, incr): return f'{path}_{incr}'
+        while Path(add_incr(path, incr)).exists(): incr += 1
+        path = add_incr(path, incr)
+    return path
```

Comment:
Adding a number to the path isnot informative. I suggest generating unique directories using timestamp as:
```python
from datetime import datetime
now = datetime.now()
timestamp = now.strftime("%Y%m%d_%H%M%S")
unique_path = f"{path}_{timestamp}"
```

Reply:
While this would make it more straightforward to make a unique path, I think this will also make it harder to identify the milestone when trying to load it. For example, if you're trying to continue from 'the second milestone' you would have to know (or look up) which timestamp this milestone has.

This might still be the best approach, or we could combine the concepts by storing milestones as {path}_{step}_{timestamp}.

Note that if we have any method for creating unique paths, we can always get the time at which that path was created from the OS.

Reply:
TODO: keep instance numbering instead of the timestamp
TODO: add docstring to _make_path_unique
TODO: rename incr to instance.

Reply:
TODO: add (str) description to config file. Not to be used by the program, but just for the researcher to describe the run or policy to test. Add comment to describe this value.